lib/hexpm_web/session.ex

```diff
@@ -1,65 +1,235 @@
+  # Serializer for Plug.Session
+  @spec encode(term()) :: String.t()
+  def encode(val), do: Jason.encode(val)
+
+  @spec decode(String.t()) :: term()
+  def decode(val), do: Jason.decode(val)
+
+  # General Public API
+  @spec get(String.t() | nil) ::
+          {:ok, Session.t()} | {:error, :expired | :no_session | :invalid_token}
+  def get(nil), do: {:error, :no_session}
+
+  def get(session_id) do
+    with {:ok, {id, token}} <- parse_session_str(session_id),
+         {:ok, session} <- get_active_session(id),
+         {:ok, session} <- validate_token(session, token) do
+      maybe_expire(session)
+    end
+  end
+
+  @spec delete(Plug.Conn.t()) :: Plug.Conn.t()
+  def delete(conn) do
+    case get_session(conn, "session_id") do
+      nil ->
+        clear_and_drop(conn)
+
+      session_id ->
+        {:ok, {id, _token}} = parse_session_str(session_id)
+
+        case get_active_session(id) do
+          {:ok, session} ->
+            expire(session)
+
+          _ ->
+            nil
+        end
+
+        clear_and_drop(conn)
+    end
+  end
+
+  @spec expire(nil | Session.t() | session_id()) :: {:ok, :expired} | {:error, :no_session}
+  def expire(nil), do: {:error, :no_session}
+
+  def expire(%Session{} = session) do
+    session
+    |> Session.expire()
+    |> Repo.delete!()
+  end
+
+  def expire(id) do
+    case Session.get_active(id) do
+      {:error, _} = err ->
+        err
+
+      {:ok, session} ->
+        Session.expire(session)
     end
   end
 
-  def put(_conn, nil, data, _opts) do
-    session = Session.build(data)
+  @spec create(Plug.Conn.t(), User.t()) :: Plug.Conn.t()
+  def create(conn, user) do
+    token = gen_token()
+
+    data = %{
+      expires_in: max_age()
+    }
 
     session =
-      if Repo.write_mode?() do
-        Repo.insert!(session)
-      else
-        Ecto.Changeset.apply_changes(session)
-      end
-
-    build_cookie(session)
-  end
-
-  def put(_conn, {id, token}, data, _opts) do
-    if Repo.write_mode?() do
-      Repo.update_all(
-        Session.by_id(id),
-        set: [
-          data: data,
-          updated_at: DateTime.utc_now()
-        ]
-      )
+      user
+      |> Session.build(data, token)
+      |> Repo.insert!()
+
+    # Prune the table when we can until a concurrent session strategy (possibly non-concurrent), 
+    # auditing capablities, and a sweeper job has been devised.
+    user
+    |> Session.all_inactive_by_user()
+    |> Repo.delete_all()
+
+    conn
+    |> configure_session(renew: true)
+    |> put_session(:session_id, to_id(session, token))
+    |> assign(:current_user, user)
+    |> assign(:current_organization, nil)
+  end
+
+  @spec max_age() :: pos_integer()
+  def max_age(), do: 60 * 60 * 24 * 30
+
+  # Private API
+  defp to_id(session, token) when byte_size(token) == 64 do
+    "#{session.uuid}#{Base.url_encode64(token)}"
+  end
+
+  defp to_id(_, _), do: {:error, :invalid_session_token}
+
+  defp gen_token(), do: :crypto.strong_rand_bytes(64)
```

Comment:
96 bytes was overkill, 32 bytes more than sufficient (256 bit of entropy is considered un-guessable iirc), 64 provides us with roughly 512 bits of entropy ( would take close to a thousands years to guess). Happy to dial it down to 32, but 64 feels soooo good 😁

Reply:
64 seems fine.